File: optimize-water-distribution-in-a-village.cpp
//Time - O(n)
//Space - O(n)
class Solution {
    
  public:
    bool static comp(vector<int> a, vector<int> b){
        return a[2]<b[2];
    }
    vector<int> parent;
    
    
    int waterDistribution(vector<int> wells, vector<vector<int>> pipes, int n){
        parent = vector<int> (n+1);
        int cost = 0;
        for(int i = 0;i<parent.size();i++){
            parent[i] = i;
        }
        
        vector<vector<int>> edges;
        for(int i = 0;i<wells.size();i++){
            edges.push_back({0,i+1,wells[i]});
        }
        
        for(int i = 0;i<pipes.size();i++){
            edges.push_back(pipes[i]);
        }
        sort(edges.begin(),edges.end(),comp);

        for(int i = 0;i<edges.size();i++){
            int x = find(edges[i][0]);
            int y = find(edges[i][1]);
            if(x!=y){
                parent[y] = x;
                cost += edges[i][2]; 
            }
        }
        return cost;
    }
    
    int find(int x){
        if(x!=parent[x]){
            parent[x] = find(parent[x]);
        }
        return parent[x];
    }
};

int main() {
    std::cout << "Hello World!\n";
    Solution sol;
    int n = 3;
    vector<int> wells {1,2,2};
    vector<vector<int>> pipes = {{1,2,1},{2,3,1}};
    
    cout<<sol.waterDistribution(wells, pipes, n);
    return 0;

}

File: optimizeWaterDistributionInAVillage.py
# Time Complexity : O(e+eloge)
# Space Complexity : O(Edges)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No


# Your code here along with comments explaining your approach

class Solution:
    def minCostToSupplyWater(self, n: int, wells: List[int], pipes: List[List[int]]) -> int:
        self.disjointSet = [0 for i in range(n+1)]
        
        res = 0
        
        for i in range(n+1):
            self.disjointSet[i] = i
        
        edges = []
        
        for i in range(1, n+1):
            edges.append([0,i, wells[i-1]])
        for pipe in pipes:
            edges.append(pipe)
            
        #sort on prices 
        edges = sorted(edges, key = lambda x:x[2])
        
        for edge in edges:
            x = self.find(edge[0])
            y = self.find(edge[1])
            if x != y:
                self.disjointSet[x] = y
                res += edge[2]
        return res
    def find(self, x):
        if  self.disjointSet[x] != x:
            self.disjointSet[x] = self.find(self.disjointSet[x])
            
        return self.disjointSet[x]